db/sync.go

```diff
@@ -51,3 +54,78 @@ func (db *DB) BeginSync(batchPushURL string, diffServerURL string, dataLayerAuth
 
 	return syncHeadRef.TargetHash(), syncInfo, nil
 }
+
+func (db *DB) MaybeEndSync(syncHead hash.Hash) (bool, []Mutation, error) {
+	v := db.Noms().ReadValue(syncHead)
+	if v == nil {
+		return false, []Mutation{}, fmt.Errorf("could not load sync head %s", syncHead)
+	}
+	var syncHeadCommit Commit
+	if err := marshal.Unmarshal(v, &syncHeadCommit); err != nil {
+		return false, []Mutation{}, err
+	}
+
+	db.lock()
+	defer db.mu.Unlock()
+
+	// Stop if someone landed a sync since this sync started.
+	syncSnapshot, err := baseSnapshot(db.noms, syncHeadCommit)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	syncSnapshotBasis, err := syncSnapshot.Basis(db.noms)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	headSnapshot, err := baseSnapshot(db.noms, db.head)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	if !syncSnapshotBasis.Original.Equals(headSnapshot.Original) {
+		return false, []Mutation{}, fmt.Errorf("sync aborted: found a newer snapshot %s on master", headSnapshot.Original.Hash())
+	}
+
+	// Determine if there are any pending mutations that we need to replay.
+	pendingCommits, err := pendingCommits(db.noms, db.head)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	commitsToReplay := filterIDsLessThanOrEqualTo(pendingCommits, syncHeadCommit.MutationID())
+	var replay []Mutation
+	if len(commitsToReplay) > 0 {
+		for _, c := range commitsToReplay {
+			var args bytes.Buffer
+			err = nomsjson.ToJSON(c.Meta.Local.Args, &args)
+			if err != nil {
+				return false, []Mutation{}, err
+			}
+			replay = append(replay, Mutation{
+				ID:   c.MutationID(),
+				Name: string(c.Meta.Local.Name),
+				Args: args.Bytes(),
+			})
+		}
+		return false, replay, nil
+	}
+
+	// TODO check invariants from synchead back to syncsnapshot.
+
+	// Sync is complete. Can't ffwd because sync head is dangling.
+	_, err = db.noms.SetHead(db.noms.GetDataset(LOCAL_DATASET), syncHeadCommit.Ref())
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	db.head = syncHeadCommit
+
+	return true, []Mutation{}, nil
+}
+
+// Assumes commits are in ascending order of mutation id.
```

Comment:
[fyi] The comment is incorrect. However, I suggest removing this function entirely (see other comment).